File: baekjoon/1012_유기농 배추/유기농 배추.js
// 실버 2 : https://www.acmicpc.net/problem/1012

const fs = require("fs");
const file = process.platform === "linux" ? "/dev/stdin" : "./text.txt";
const input = fs.readFileSync(file).toString().trim().split("\n");

const t = +input.shift();

const inputArr = input.map((v) => v.split(" ").map(Number));

const dx = [-1, 0, 1, 0];
const dy = [0, 1, 0, -1];

const dfs = (x, y, n, m, graph) => {
  let nx = 0;
  let ny = 0;
  for (let i = 0; i < 4; i++) {
    nx = x + dx[i];
    ny = y + dy[i];
    if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
      if (graph[nx][ny] == 1) {
        graph[nx][ny] = -1;
        dfs(nx, ny, n, m, graph);
      }
    }
  }
};

for (let i = 0; i < t; i++) {
  let [m, n, k] = inputArr.shift();
  let graph = Array.from({ length: n }, () => Array(m).fill(0));
  while (k > 0) {
    k--;
    const [y, x] = inputArr.shift();
    graph[x][y] = 1;
  }
  let result = 0;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (graph[i][j] === 1) {
        dfs(i, j, n, m, graph);
        result += 1;
      }
    }
  }
  console.log(result);
}

File: baekjoon/1012_유기농 배추/유기농 배추.py
# 실버 2 : https://www.acmicpc.net/problem/1012

# 파이썬의 재귀 호출 한도를 설정 => 없으면 런타임 에러 (RecursionError) 발생!!!
import sys 
sys.setrecursionlimit(10000) 

t = int(input())

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

def dfs(x, y):
  for i in range(4):
    nx = x + dx[i]
    ny = y + dy[i]
    if (nx >= 0 and ny >= 0 and nx < n and ny < m):
      if (graph[nx][ny] == 1):
        graph[nx][ny] = -1 # 방문 처리를 visited 배열 생성하지 말고 그냥 배열 -1 값으로 초기화
        dfs(nx, ny)

for _ in range(t):
  m, n, k = map(int, input().split())
  graph = [[0 for _ in range(m)] for _ in range(n)]
  for _ in range(k):
    y, x = map(int, input().split())
    graph[x][y] = 1
  result = 0
  for i in range(n):
    for j in range(m):
      if graph[i][j] == 1:
        dfs(i, j)
        result += 1
  print(result)
